pages/studentAssessmentExam/studentAssessmentExam.js

```diff
@@ -33,9 +56,127 @@ router.post('/', function(req, res, next) {
     if (res.locals.assessment.type !== 'Exam') return next();
     if (!res.locals.authz_result.authorized_edit) return next(error.make(403, 'Not authorized', res.locals));
     if (req.body.__action == 'newInstance') {
-        assessment.makeAssessmentInstance(res.locals.assessment.id, res.locals.user.user_id, res.locals.assessment.group_work, res.locals.authn_user.user_id, res.locals.authz_data.mode, res.locals.authz_result.time_limit_min, res.locals.req_date, (err, assessment_instance_id) => {
+        //make sure there is no new instance created by student's teammate simultaneously
```

Comment:
There is a possible race condition here:

* Student A calls `select_single_assessment_instance` and finds nothing.
* Student B calls `select_single_assessment_instance` and finds nothing.
* Student A calls `assessment.makeAssessmentInstance`.
* Student B calls `assessment.makeAssessmentInstance` => ERROR.

This will definitely happen in practice.

One way to avoid this condition might be to both check for and make an assessment instance in the same SQL query.

Reply:
Thank you! I just moved this race condition check into assessment.makeAssessmentInstance